programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+        ret = MBEDTLS_ERR_SERIALIZE_BAD_INPUT;                          \
+        break;                                                          \
+    } else {}
+#define CHECK_LENGTH( i, n )                                            \
+    if( inputs[i]->size < ( n ) )                                       \
+    {                                                                   \
+        DBG( "parameter %u too short: %zu < %zu", (unsigned) i, inputs[i]->size, (size_t) ( n ) ); \
+        ret = MBEDTLS_ERR_SERIALIZE_BAD_INPUT;                          \
+        break;                                                          \
+    } else {}
+#define ALLOC_OUTPUT( i, length ) \
+    if( ( outputs[i] = alloc_item( length ) ) == NULL )                 \
+    {                                                                   \
+        DBG( "failed to allocate %zu bytes for output %u", (size_t) ( length ), (unsigned) ( i ) );   \
+        ret = MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED;                       \
+        break;                                                          \
+    } else {}
+
+
+/** File context 
+ */
+typedef struct {
+    void *      file;
+    int8_t      inUse;
+} mbedtls_serialize_file_context_t;
+
+#define MBEDTLS_SERIALIZE_MAX_FILES     100
+static mbedtls_serialize_file_context_t files[MBEDTLS_SERIALIZE_MAX_FILES];
+
+int32_t alloc_file_context()
+{
+    int i, file_id = -1;
+
+    for ( i = 0; i < MBEDTLS_SERIALIZE_MAX_FILES; i++ )
+    {
+        if ( !files[i].inUse )
+        {
+            files[i].inUse = 1;
+            file_id = i + 1; // file Id can not be 0 i.e. same as NULL
+            break;
+        }
+    }
+    return file_id;
+}
+
+int free_file_context( int32_t file_id )
+{
+    int ret = -1;
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES )
+    {
+        files[file_id].inUse = 0;
+        if ( files[file_id].file )
+        {
+            files[file_id].file = NULL;
+            ret = 0;
+        }
+    }
+    return ret;
+}
+
+void * get_file_from_id( int32_t file_id )
+{
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES && files[file_id].inUse )
+    {
+        return files[file_id].file;
+    }
+    DBG( "Failed to find file for Index %d", file_id );
+    return NULL;
+}
+
+/** Execute an offloaded function.
+ *
+ * @param ctx       Offloading contecxt
+ * @param function  Function to execute (MBEDTLS_SERIALIZE_FUNCTION_xxx)
+ * @param outputs   Array[16] of outputs. Initially all-null. On return,
+ *                  for a function with N outputs, positions 0 to N-1 will
+ *                  be filled with output parameters allocated with alloc_item.
+ * @return status (0 for ok, MBEDTLS_ERR_xxx on error)
+ */
+static uint32_t mbedtls_serialize_perform( mbedtls_serialize_context_t *ctx_p,
+                                           uint32_t function,
+                                           mbedtls_serialize_item_t **outputs )
```

Comment:
I think they can be arrays, and despite the fact that (except of the use in `sizeof`) they behave the same as the pointers to their first element, the reason to suggest it was that it might ease reading the code.